### PR TITLE
[misc] fix: Improve compare.py robustness for multi-GPU and vocab-padded models

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary
- Fix NCCL broadcast dtype mismatch by converting HF logits to `float32` before the fallback tensor creation path, ensuring all ranks use the same dtype
- Handle Megatron vocab-size padding by truncating logits to HF vocab size before comparison, eliminating the shape-mismatch code path
- Simplify `vlm_forward_step` return, remove unused `gc` import, `grad_scale_func` workaround, and unnecessary EP rank guard
- Add `torch.distributed.barrier()` after HF results broadcast for synchronization safety

## Test plan
- [x] Verified on remote cluster with `Qwen/Qwen3-0.6B`: token match ✅, cosine similarity 99.99%
- [ ] CI passes

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **Refactor**
  * Enhanced model comparison utility to improve accuracy and consistency across distributed training scenarios. Simplified forward-output handling and synchronized results across compute ranks. Updated comparison metrics calculation for more reliable model alignment validation.

### CODE REVIEW

**Review by cuichenx — DISMISSED** [review]
LGTM!

File: examples/conversion/compare_hf_and_megatron/compare.py [hunk 1]
```diff
@@ -792,10 +784,7 @@ def compare_models_one_step(args) -> None:
             megatron_logits = megatron_output[0, -1, :]
             megatron_next_token = torch.argmax(megatron_logits, dim=-1)
 
-            if not torch.distributed.is_initialized() or (
-                parallel_state.get_tensor_model_parallel_rank() == 0
-                and parallel_state.get_expert_model_parallel_rank() == 0
-            ):
+            if not torch.distributed.is_initialized() or parallel_state.get_tensor_model_parallel_rank() == 0:
```
Comment (cuichenx) [inline examples/conversion/compare_hf_and_megatron/compare.py]:
```suggestion
            if not torch.distributed.is_initialized() or (parallel_state.get_tensor_model_parallel_rank() == 0 and parallel_state.get_expert_model_parallel_rank() == 0):
```
a small suggestion for EP>1 runs so that output is not garbled

File: examples/conversion/compare_hf_and_megatron/compare.py [hunk 2]
```diff
@@ -616,11 +609,8 @@ def _load_megatron_model(args):
         model_provider.finalize()
         megatron_model = model_provider.provide_distributed_model(wrap_with_ddp=False)
 
-    # Workaround: disable MTP for inference (causes hangs on NCCL collectives)
     for m in megatron_model:
```
Comment (yaoyu-33) [inline examples/conversion/compare_hf_and_megatron/compare.py:612]: check _disable_mtp util

**Comment by yaoyu-33** [comment]
covered by https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2602
